shedlock-provider-jedis/src/main/java/net/javacrumbs/shedlock/provider/jedis/JedisLockProvider.java

```diff
@@ -0,0 +1,133 @@
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+
+/**
+ * Uses Redis's `SET resource-name anystring NX PX max-lock-ms-time` as locking mechanism.
+ *
+ * See https://redis.io/commands/set
+ *
+ */
+public class JedisLockProvider implements LockProvider {
+
+    private static final String KEY_PREFIX = "job-lock";
+    private static final String ENV_DEFAULT = "default";
+
+    private static final long AT_MOST_HOURS_DEFAULT = 1;
+
+    // Redis Flags
+    private static final String SET_IF_NOT_EXIST = "NX";
+    private static final String SET_EXPIRE_TIME_IN_MS = "PX";
+
+    private JedisPool jedisPool;
+    private String environment;
+    private Duration atMostHoursDefault;
+
+    public JedisLockProvider(JedisPool jedisPool) {
+        this(jedisPool, ENV_DEFAULT, AT_MOST_HOURS_DEFAULT);
+    }
+
+    public JedisLockProvider(JedisPool jedisPool, String environment) {
+        this(jedisPool, environment, AT_MOST_HOURS_DEFAULT);
+    }
+
+    public JedisLockProvider(JedisPool jedisPool, String environment, long atMostHoursDefault) {
+        this.jedisPool = jedisPool;
+        this.environment = environment;
+        this.atMostHoursDefault = Duration.of(atMostHoursDefault, ChronoUnit.HOURS);
+    }
+
+    @Override
+    public Optional<SimpleLock> lock(LockConfiguration lockConfiguration) {
+        Instant now = Instant.now();
+
+        // Get 'furthest out' configured expire time for lock TTL
+        long expireTime = Math.max(
+            Duration.between(now, lockConfiguration.getLockAtMostUntil()).toMillis(),
+            Duration.between(now, lockConfiguration.getLockAtLeastUntil()).toMillis());
+
+        String key = buildKey(lockConfiguration.getName(), this.environment);
+
+        try (Jedis jedis = jedisPool.getResource()) {
+            String rez = jedis.set(key,
+                    buildValue(),
+                    SET_IF_NOT_EXIST,
+                    SET_EXPIRE_TIME_IN_MS,
+                    expireTime < 0 ? atMostHoursDefault.toMillis() : expireTime); // And if not set, use default
+
+            if (rez != null && "OK".equals(rez)) {
+                return Optional.of(new RedisLock(key, jedisPool, lockConfiguration));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static final class RedisLock implements SimpleLock {
+        private final String key;
+        private final JedisPool jedisPool;
+        private final LockConfiguration lockConfiguration;
+
+        private RedisLock(String key, JedisPool jedisPool, LockConfiguration lockConfiguration) {
+            this.key = key;
+            this.jedisPool = jedisPool;
+            this.lockConfiguration = lockConfiguration;
+        }
+
+        @Override
+        public void unlock() {
+            Instant now = Instant.now();
+            Instant atLeastUntil = lockConfiguration.getLockAtLeastUntil();
+
+            if (now.equals(atLeastUntil) || now.isAfter(atLeastUntil)) {
```

Comment:
Let's imagine I have lockAtleastFor 5 minutes and lockAtMostFor 1 hour and the task takes 2 minutes. We should hold the lock for 3 more minutes.